### PR TITLE
New Python Solutions: USACO 2019 Dec Bronze Q1, Feb 2020 Bronze Q1

### PR DESCRIPTION
**Pull request by AAkashyap1**

### CODE REVIEW

File: solutions/usaco-963.mdx [hunk 1]
```diff
@@ -133,3 +133,43 @@ int main(){
...
+
+bad_pairs = 0
+
+# Find the number of pairs of cows which appear out of order in any practice session
+for i in range(n - 1):
+  for j in range(i + 1, n):
+    for l in range(1, k):
+      if i == 0 and j == 1:
+        print(cows[l].index(cows[0][j]))
+        print(cows[l].index(cows[0][i]))
```
Comment (thecodingwizard) [inline solutions/usaco-963.mdx:163]: wait what's the point of these two lines?

Comment (AAkashyap1) [reply solutions/usaco-963.mdx:163]: oof gotta fix that

**Comment by thecodingwizard** [comment]
For the gymnastics problem, I think it might be better if the solution was the same as the internal solution rather than being slightly modified from the internal solution, since this is a Bronze problem and the difficulty is implementation, not logic, and having slightly different solutions is more confusing than the same solution across all languages.

For example, the official editorial checks whether A is consistent with B by checking to see if the number of practice sessions where A is ranked before B is equal to K, while your implementation checks it in a different way. I think it would be less confusing if both solutions checked it in the same way.

Thanks for contributing Python solutions though :)

**Review by MrinallU — APPROVED** [review]

File: solutions/usaco-1011.mdx [hunk 2]
```diff
@@ -63,3 +63,40 @@ public class triangles{
...
+        for k in range(n):
+          if k != j and k != i:
+            if vertices[i][0] == vertices[k][0]:
+              area = abs(vertices[k][1]-vertices[i][1])*abs(vertices[j][0]-vertices[i][0])
+              max_area = max(max_area, area)
+      elif vertices[j][0] == vertices[i][0]:
+        for k in range(n):
+          if k != j and k != i:
+            if vertices[i][1] == vertices[k][1]:
+              area = abs(vertices[k][0]-vertices[i][0])*abs(vertices[j][1]-vertices[i][1])
```
Comment (maggie-j-liu) [inline solutions/usaco-1011.mdx:93]:
```suggestion
              area = abs(vertices[k][1] - vertices[i][1]) * abs(vertices[j][0] - vertices[i][0])
              max_area = max(max_area, area)
      elif vertices[j][0] == vertices[i][0]:
        for k in range(n):
          if k != j and k != i:
            if vertices[i][1] == vertices[k][1]:
              area = abs(vertices[k][0] - vertices[i][0]) * abs(vertices[j][1] - vertices[i][1])
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-1011.mdx [hunk 3]
```diff
@@ -63,3 +63,39 @@ public class triangles{
...
+			if vertices[j][1] == vertices[i][1]:
+				for k in range(n):
+					if k != j and k != i:
+						if vertices[i][0] == vertices[k][0]:
+							area = abs(vertices[k][1] - vertices[i][1]) * abs(vertices[j][0] - vertices[i][0])
+							max_area = max(max_area, area)
+			elif vertices[j][0] == vertices[i][0]:
+				for k in range(n):
+					if k != j and k != i:
+						if vertices[i][1] == vertices[k][1]:
```
Comment (maggie-j-liu) [inline solutions/usaco-1011.mdx:92]: maybe combine some of the if statements?

File: solutions/usaco-963.mdx [hunk 4]
```diff
@@ -133,3 +133,40 @@ int main(){
...
+## Official Solution with Slight Modification
+
+As in the official solution, we can go through every pair of cows in the first practice session, and determine whether or not they appear in the same order for every other practice session. We can create a value storing the number of pairs that do not fit this constraint. Then we can subtact this value from the total number of pairs, thus giving us the number of consistent pairs. 
+
+**Time Complexity:** $\mathcal{O}(KN^2)$
+
+<LanguageSection>
+<PySection>
+
+```py
```
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/usaco-963.mdx:146]: > use tabs instead of spaces to indent

not fixed yet

File: solutions/usaco-1011.mdx [hunk 5]
```diff
@@ -63,3 +63,39 @@ public class triangles{
 </JavaSection>
 
 </LanguageSection>
+
+<LanguageSection>
+ 
+<PySection>
```
Comment (maggie-j-liu) [inline solutions/usaco-1011.mdx:69]: move this pysection into the languagesection above. you only need a new languagesection if it's a different method

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Comment by dongliu0426** [comment]
https://github.com/cpinitiative/usaco-guide/pull/1592/files#r645833512

File: solutions/usaco-1011.mdx (hunk 3 above)
Comment (AAkashyap1) [reply solutions/usaco-1011.mdx]: How does that look?

**Comment by AAkashyap1** [comment]
> https://github.com/cpinitiative/usaco-guide/pull/1592/files#r645833512
Just fixed this

File: solutions/usaco-1011.mdx (hunk 3 above)
Comment (maggie-j-liu) [reply solutions/usaco-1011.mdx]: I think you put two tabs for each indent?

Comment (AAkashyap1) [reply solutions/usaco-1011.mdx]: Does it look food now?

Comment (maggie-j-liu) [reply solutions/usaco-1011.mdx]: Yep!

File: solutions/usaco-963.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/usaco-963.mdx:146]: i think you changed it to tabs, but some are still spaces? could you check this again?

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-963.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (AAkashyap1) [reply solutions/usaco-963.mdx:146]: Ok it should finally be good i think

File: solutions/usaco-963.mdx [hunk 6]
```diff
@@ -133,3 +133,39 @@ int main(){
...
+
+# Find the number of pairs of cows which appear out of order in any practice session
+for i in range(n - 1):
+	for j in range(i + 1, n):
+		for l in range(1, k):
+			if cows[l].index(cows[0][j]) < cows[l].index(cows[0][i]):
+				bad_pairs += 1
+				break
+
+ans = n * (n - 1) / 2 - bad_pairs
```
Comment (dongliu0426) [inline solutions/usaco-963.mdx:165]: better practice to use `//`

File: solutions/usaco-963.mdx [hunk 7]
```diff
@@ -133,3 +133,39 @@ int main(){
...
+<PySection>
+
+```py
+file_in = open('gymnastics.in')
+data = file_in.read().strip().split('\n')
+k, n = map(int, data[0].split())
+
+cows = [list(map(int, data[i + 1].split())) for i in range(k)]
+
+print(cows)
```
Comment (dongliu0426) [inline solutions/usaco-963.mdx:153]: remove this (?)

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/usaco-963.mdx [hunk 8]
```diff
@@ -132,4 +132,32 @@ int main(){
...
+<PySection>
+
+```py
+file_in = open('gymnastics.in')
+data = file_in.read().strip().split('\n')
+k, n = map(int, data[0].split())
+
+cows = [list(map(int, data[i + 1].split())) for i in range(k)]
+
+bad_pairs = 0
```
Comment (maggie-j-liu) [inline solutions/usaco-963.mdx:145]: can remove this
```suggestion
```

**Review by maggie-j-liu — APPROVED** [review]

**Review by jessechoe10 — APPROVED** [review]

**Review by dongliu0426 — APPROVED** [review]

File: solutions/usaco-1011.mdx [hunk 9]
```diff
@@ -61,5 +61,35 @@ public class triangles{
...
+		if i != j and vertices[j][1] == vertices[i][1]:
+			for k in range(n):
+				if (k != j and k != i) and vertices[i][0] == vertices[k][0]:
+					area = abs(vertices[k][1] - vertices[i][1]) * abs(vertices[j][0] - vertices[i][0])
+					max_area = max(max_area, area)
+		elif i != j and vertices[j][0] == vertices[i][0]:
+			for k in range(n):
+				if (k != j and k != i) and vertices[i][1] == vertices[k][1]:
+					area = abs(vertices[k][0] - vertices[i][0]) * abs(vertices[j][1] - vertices[i][1])
+					max_area = max(max_area, area)
```
Comment (frodakcin) [inline solutions/usaco-1011.mdx:87]: Redundant; code passes fine without these lines ( only assuming i and j horizontal )

**Review by frodakcin — CHANGES_REQUESTED** [review]

**Review by MrinallU — APPROVED** [review]

File: solutions/usaco-1011.mdx (hunk 9 above)
Comment (AAkashyap1) [reply solutions/usaco-1011.mdx:87]: Wait so I should just delete lines 83-87? Will the code still work perfectly?

Comment (dongliu0426) [reply solutions/usaco-1011.mdx:87]: > ( only assuming i and j horizontal )

Comment (AAkashyap1) [reply solutions/usaco-1011.mdx:87]: But what if i and j aren't horizontal, then won't the code fail?

Comment (dongliu0426) [reply solutions/usaco-1011.mdx:87]: then loop through all pairs, (i, j) such that they are horizontal (?)

Comment (AAkashyap1) [reply solutions/usaco-1011.mdx:87]: But how would i account for the cases when they aren't

Comment (dongliu0426) [reply solutions/usaco-1011.mdx:87]: loop through all pairs and check if they're horizontal...

File: solutions/usaco-963.mdx [hunk 10]
```diff
@@ -132,4 +132,31 @@ int main(){
...
+	for j in range(i + 1, n):
+		count = 0
+		for l in range(1, k):
+			if cows[l].index(cows[0][j]) > cows[l].index(cows[0][i]):
+				count += 1
+		if count == k - 1:
+			ans += 1
+
+
+print(str(ans), file=open('gymnastics.out', 'w'))
```
Comment (SansPapyrus683) [inline solutions/usaco-963.mdx:158]: no need for the `str`

File: solutions/usaco-1011.mdx [hunk 11]
```diff
@@ -61,5 +61,35 @@ public class triangles{
...
+
+print(str(max_area), file=open('triangles.out', 'w'))
```
Comment (SansPapyrus683) [inline solutions/usaco-1011.mdx:89]: str not needed here either

File: solutions/usaco-963.mdx [hunk 12]
```diff
@@ -132,4 +132,31 @@ int main(){
...
+```
```
Comment (SansPapyrus683) [inline solutions/usaco-963.mdx:159]: lines 90, 144, 146, and 156 are unnecessary whitespace

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
nitpicking time

**Comment by dongliu0426** [comment]
@AAkashyap1 Could you address the requested changes?

File: solutions/usaco-1011.mdx (hunk 9 above)
Comment (AAkashyap1) [reply solutions/usaco-1011.mdx:87]: Wait but then what happens if they arent (sry for bothering)

Comment (frodakcin) [reply solutions/usaco-1011.mdx:87]: Yes you delete the lines and the code will work perfectly.

Comment (frodakcin) [reply solutions/usaco-1011.mdx:87]: > Wait but then what happens if they arent

Every triangle that you consider has a horizontal edge. This is the (i, j) edge -- it will always exist

Comment (dongliu0426) [reply solutions/usaco-1011.mdx:87]: @AAkashyap1 ^^

**Comment by dongliu0426** [comment]
Changes requested have not been made. Free free to create a new PR.
